contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,204 @@
+# Encryption
+
+## Abstract
+
+The scope of this proposal is to ensure that resources can be encrypted at the datastore layer. Encrypting data at REST and via third party vendors is a desired feature but outside the scope of this proposal. Encryption will be optional for any resource but we suspect it will be used for the Secret resource in most cases.
+
+Allowing sensitive data to be encrypted adheres to best practices as well as other requirements such as HIPAA.
+
+How encryption keys are delivered to the machine running the Kubernetes apiserver is of relevance - we assume that the encryption at rest pattern is loosely coupled to how those keys are delivered and secured on disk.
+
+In general, full disk encryption of the volumes storing etcd data is generally preferred - this proposal focuses on scenarios where additional protection is desired against malicious parties gaining read access to the etcd API or a running etcd instance without access to memory of the etcd process.
+
+## High level design
+Before a resource is written to etcd and before it is read, an encryption provider will take the plaintext data and encrypt it.
+These providers will be able to be created and turned on depending on the users needs or requirements and will adhere to an encryption interface.
+This interface will provide the abstraction to allow various encryption mechanisms to be implemented, as well as for the method of encryption to be rotated over time.
```

Comment:
Automated key rotation might be desirable but IMO not necessary. What we did in swarm was piggyback on transitions from an "unlocked" cluster to a "locked" cluster, rotating both the KEK and the DEK (and doing a snapshot w/ the new DEK). This allows users to force a full rotation of keys/re-encryption of data, as a byproduct of the operation that will need it the most (transitioning from a state where a KEK is on disk to a state where the KEK only lives in memory)